### ISSUE TITLE
Add IPv6 support for RPC server

### ISSUE THREAD

**Issue by slokhorst** [transmission/transmission#59]
This is a follow up on [Trac#2236](https://trac.transmissionbt.com/ticket/2236).

In the Trac discussion, it is mentioned that IPv6 support requires changes to libevent. I decided to check this, but it turns out that **IPv6 works fine with the current libevent**.

You can verify this yourself by applying [this dirty hack](https://gist.github.com/slokhorst/ce498cca9c6354080e9e49f7057ad3e6), which disables IPv4 and listens to IPv6 instead. [Proof that it works](http://imgur.com/a/ZdLrF).

Properly supporting IPv6 requires some changes to `libtransmission/rpc-server.c`, and maybe the addition of an additional configuration variable (`rpc-bind-address-{ipv4,ipv6}` instead of the current `rpc-bind-address`). I'm not familiar with the Transmission code at all, but if no one else is interested, I will attempt to create a patch myself.

**Comment by slokhorst** [comment]
I'm working on a patch, but I've run into a problem when binding on both `0.0.0.0` and `::` ([libevent#394](https://github.com/libevent/libevent/issues/394)). Any help would be appreciated.

**Comment by nschloe** [comment]
Perhaps this would generate a little more attention if submitted as a PR (with the above remarks)?

**Comment by slokhorst** [comment]
Yes, I'll probably do that.

The only "solution" I can think of is to always bind to all addresses (`::` and `0.0.0.0`). I'm not sure if this will be a problem. Security-wise I think it should be OK, since you need to manually add hosts to the RPC whitelist anyway.

**Comment by slokhorst** [comment]
@nschloe maybe want to test #161? If more people confirm it to be working we can hopefully get it merged soon.

**Comment by nschloe** [comment]
Does anyone happen to know where to place the config file when installing from master? Its previous location `/etc/transmission-daemon/settings.json` seems to be disregarded.

**Comment by slokhorst** [comment]
It should look at `~/.config/transmission-daemon/settings.json` first.

**Comment by nschloe** [comment]
I can reach the web server with IPv4 alright and ping6 the machine too. However, whenever I start the daemon, the `rpc-bind-adress` is set to `0.0.0.0`:
```
$ grep bind ~/.config/transmission-daemon/settings.json 
    "bind-address-ipv4": "0.0.0.0",
    "bind-address-ipv6": "::",
    "rpc-bind-address": "::",
$ /usr/local/bin/transmission-daemon
$ grep bind ~/.config/transmission-daemon/settings.json 
    "bind-address-ipv4": "0.0.0.0",
    "bind-address-ipv6": "::",
    "rpc-bind-address": "0.0.0.0",
```
Any hints?

**Comment by slokhorst** [comment]
You need to first shut down the daemon (`killall transmission-daemon`), then edit the file, then start the daemon (I recommend with the `-f` flag, so it stays in the foreground, you can see any errors, and you can easily kill it).

**Comment by nschloe** [comment]
> recommend with the -f flag, so it stays in the foreground

Great tip.

I'm doing just that and other config options, like to whitelist setting, are preserved, too. Only `rpc-bind-address` seems to change. Perhaps this
```
RPC Server :: is not an IPv4 address. RPC listeners must be IPv4
```
is an issue?

Can you see anything suspicious in the output?
```
[2017-02-23 13:39:22.863] Transmission 2.92+ (0f75e64d71) started (/home/pi/software/transmission/source-upstream/libtransmission/session.c:739)
[2017-02-23 13:39:22.864] RPC Server Adding address to whitelist: 127.0.0.1 (/home/pi/software/transmission/source-upstream/libtransmission/rpc-server.c:879)
[2017-02-23 13:39:22.864] RPC Server Adding address to whitelist: ::1 (/home/pi/software/transmission/source-upstream/libtransmission/rpc-server.c:879)
[2017-02-23 13:39:22.864] RPC Server :: is not an IPv4 address. RPC listeners must be IPv4 (/home/pi/software/transmission/source-upstream/libtransmission/rpc-server.c:1078)
[2017-02-23 13:39:22.864] RPC Server Serving RPC and Web requests on port 127.0.0.1:9091/transmission/ (/home/pi/software/transmission/source-upstream/libtransmission/rpc-server.c:1085)
[2017-02-23 13:39:22.864] UDP Failed to set receive buffer: requested 4194304, got 327680 (/home/pi/software/transmission/source-upstream/libtransmission/tr-udp.c:84)
[2017-02-23 13:39:22.864] UDP Please add the line "net.core.rmem_max = 4194304" to /etc/sysctl.conf (/home/pi/software/transmission/source-upstream/libtransmission/tr-udp.c:89)
[2017-02-23 13:39:22.864] UDP Failed to set send buffer: requested 1048576, got 327680 (/home/pi/software/transmission/source-upstream/libtransmission/tr-udp.c:95)
[2017-02-23 13:39:22.864] UDP Please add the line "net.core.wmem_max = 1048576" to /etc/sysctl.conf (/home/pi/software/transmission/source-upstream/libtransmission/tr-udp.c:100)
[2017-02-23 13:39:22.864] UDP Failed to set receive buffer: requested 4194304, got 327680 (/home/pi/software/transmission/source-upstream/libtransmission/tr-udp.c:84)
[2017-02-23 13:39:22.864] UDP Please add the line "net.core.rmem_max = 4194304" to /etc/sysctl.conf (/home/pi/software/transmission/source-upstream/libtransmission/tr-udp.c:89)
[2017-02-23 13:39:22.864] UDP Failed to set send buffer: requested 1048576, got 327680 (/home/pi/software/transmission/source-upstream/libtransmission/tr-udp.c:95)
[2017-02-23 13:39:22.864] UDP Please add the line "net.core.wmem_max = 1048576" to /etc/sysctl.conf (/home/pi/software/transmission/source-upstream/libtransmission/tr-udp.c:100)
[2017-02-23 13:39:22.864] DHT Reusing old id (/home/pi/software/transmission/source-upstream/libtransmission/tr-dht.c:306)
[2017-02-23 13:39:22.864] DHT Bootstrapping from 60 IPv4 nodes (/home/pi/software/transmission/source-upstream/libtransmission/tr-dht.c:155)
[2017-02-23 13:39:22.864] Using settings from "/home/pi/.config/transmission-daemon" (/home/pi/software/transmission/source-upstream/daemon/daemon.c:527)
[2017-02-23 13:39:22.864] Saved "/home/pi/.config/transmission-daemon/settings.json" (/home/pi/software/transmission/source-upstream/libtransmission/variant.c:1265)
[2017-02-23 13:39:22.864] Port Forwarding (NAT-PMP) initnatpmp succeeded (0) (/home/pi/software/transmission/source-upstream/libtransmission/natpmp.c:69)
[2017-02-23 13:39:22.864] Port Forwarding (NAT-PMP) sendpublicaddressrequest succeeded (2) (/home/pi/software/transmission/source-upstream/libtransmission/natpmp.c:69)
[2017-02-23 13:39:28.864] Port Forwarding State changed from "Not forwarded" to "Starting" (/home/pi/software/transmission/source-upstream/libtransmission/port-forwarding.c:91)
[2017-02-23 13:39:29.864] Port Forwarding State changed from "Starting" to "???" (/home/pi/software/transmission/source-upstream/libtransmission/port-forwarding.c:91)
```

**Comment by slokhorst** [comment]
Are you using the code from my pull request? You can get it like so:
```bash
https://github.com/transmission/transmission.git #probably already did that
git fetch origin pull/161/head:pr161
git checkout pr161
```

**Comment by nschloe** [comment]
Success!

**Comment by nschloe** [comment]
Let me try again with the patch rebased onto master though. (That didn't work earlier.)

**Comment by nschloe** [comment]
Nope, everything alright it seems. No idea what went wrong earlier.

**Comment by slokhorst** [comment]
Ok, thanks for testing! I've rebased onto master to be sure.

**Comment by darkrain88** [comment]
does 2.94 support to bind rpc ipv6?

config to :: it seems no work.

**Comment by darkrain88** [comment]
> Success!

i stop the daemon ,and edit the setting.jason  rpc bind to :: and then restart, the config file return to 0.0.0. again.

my version is 2.94 on Openwrt.

how to make it work, thanks

**Comment by slokhorst** [comment]
@darkrain88 no, 2.94 does not support it unfortunately.

You can manually patch 2.94 (not sure how that would work with OpenWRT), or have to wait for 3.00 to be released (I'm afraid that it could be a while until that happens as there's not much activity here).

**Comment by darkrain88** [comment]
@slokhorst  thanks

**Comment by 1265578519** [comment]
是需要打上面发的补丁才支持ipv6吗

**Comment by newcoderlife** [comment]
Yes. It will not be released until version 3.

**Comment by Duffyx** [comment]
Ok, so I'm a bit confused, does this change make it possible to configure the RPC on both an IPv4 and IPv6 address, or is it mutually exclusive?

**Comment by blacklion** [comment]
"::" doesn't work as IPv4 on FreeBSD.
So, it is impossible to make RPC listen on both IPv4 and IPv6 on FreeBSD now :-(

**Comment by Duffyx** [comment]
> "::" doesn't work as IPv4 on FreeBSD.
> So, it is impossible to make RPC listen on both IPv4 and IPv6 on FreeBSD now :-(

See also: https://github.com/transmission/transmission/pull/161
